02 SFC Properties/README.md

```diff
@@ -0,0 +1,333 @@
++    <input
++      v-bind:value="message"
++      v-on:input="message = $event.target.value"
++    />
++  </div>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+
+export default Vue.extend({
+  name: 'App',
+  data() {
+    return {
+      message: 'Hello from Vue.js',
+    };
+  }
+};
+</script>
+
+```
+
+- We can use too the shorthands:
+
+### ./src/App.vue
+
+```diff
+<template>
+  <div>
+    <h1>{{message}}</h1>
+    <input
+-      v-bind:value="message"
++      :value="message"
+-      v-on:input="message = $event.target.value"
++      @input="message = $event.target.value"
+    />
+  </div>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+
+export default Vue.extend({
+  name: 'App',
+  data() {
+    return {
+      message: 'Hello from Vue.js',
+    };
+  }
+};
+</script>
+
+```
+
+- And finally, use method instead of inline function:
+
+### ./src/App.vue
+
+```diff
+<template>
+  <div>
+    <h1>{{message}}</h1>
+    <input
+      :value="message"
+-      @input="message = $event.target.value"
++      @input="onChange"
+    />
+  </div>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+
+export default Vue.extend({
+  name: 'App',
+  data() {
+    return {
+      message: 'Hello from Vue.js',
+    };
+  },
++ methods: {
++   onChange(event) {
++     this.message = event.target.value;
++   },
++ },  
+};
+</script>
+
+```
+
+> Reference: [_Don't use arrow functions_](https://vuejs.org/v2/guide/instance.html#Instance-Lifecycle-Hooks)
+
+- Using syntactic sugar `v-model`:
+
+### ./src/App.vue
+
+```diff
+<template>
+  <div>
+    <h1>{{message}}</h1>
+    <input
+-       :value="message"
+-       @input="onChange"
++       v-model="message"
+    />
+  </div>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+
+export default Vue.extend({
+  name: 'App',
+  data() {
+    return {
+      message: 'Hello from Vue.js',
+    };
+  },
+- methods: {
+-   onChange(event) {
+-     this.message = event.target.value;
+-   },
+- },
+};
+</script>
+
+```
+
+- Create our first component `Hello.vue`:
+
+### ./src/Hello.vue
+
+```javascript
+<template>
+  <input
+    v-model="value"
+  />
+</template>
+
+<script>
+export default {
+  name: 'hello',
+  props: {
+    value: String,
+  },
+};
+</script>
+
+```
+
+- Update `App.vue`:
+
+### ./src/App.vue
+
+```diff
+<template>
+  <div>
+    <h1>{{message}}</h1>
+-    <input
++    <hello-component
+      v-model="message"
+    />
+  </div>
+</template>
+
+<script>
++ import HelloComponent from './Hello.vue';
+
+export default {
+  name: 'App',
++ components: {
++   HelloComponent,
++ },
+  data() {
+    return {
+      message: 'Hello from Vue.js',
+    };
+  },
+};
+</script>
+
+```
+
+Vue
+
+- Since we are using `v-model` in `HelloComponent` we are mutating `value` prop and it's forbidden. To solve this:
+
+### ./src/Hello.vue
+
+```diff
+<template>
+  <input
+-     v-model="value"
++     :value="value"
++     @input="onChange"
+  />
+</template>
+
+<script>
```

Comment:
Use Vue.extend to avoid typescript error not found $emit:

```
<script lang="ts">
import Vue from "vue";
export default Vue.extend({
  name: "hello",
  props: {
    value: String
  },
  methods: {
    onChange: function(event) {
      this.$emit("input", event.target.value);
    }
  }
});
</script>
```